gibbon-compiler/src/Gibbon/Passes/MarkTailCalls.hs

```diff
@@ -0,0 +1,154 @@
+module Gibbon.Passes.MarkTailCalls (markTailCalls) where 
+
+
+import qualified Data.List as L
+import Data.Maybe ( fromJust )
+import qualified Data.Map as M
+import qualified Data.Set as S
+import Data.Foldable ( foldrM )
+import Prelude as P
+
+import Gibbon.Common
+import Gibbon.DynFlags
+import Gibbon.NewL2.Syntax as NewL2
+import Gibbon.L2.Syntax as Old
+
+markTailCalls :: NewL2.Prog2 -> PassM NewL2.Prog2
+markTailCalls Prog{ddefs,fundefs,mainExp} = do
+   fds' <- mapM (markTailCallsFn ddefs) $ M.elems fundefs
+   let newFundefs = M.fromList $ map (\f -> (funName f,f)) fds'
+   let newProg = Prog{ddefs=ddefs, fundefs=newFundefs, mainExp=mainExp}
+   pure $ newProg {- dbgTraceIt (sdoc newProg) dbgTraceIt (sdoc $ M.elems fundefs')-}
+
+
+markTailCallsFn :: NewL2.DDefs2 -> NewL2.FunDef2 -> PassM NewL2.FunDef2
+markTailCallsFn ddefs f@FunDef{funName, funArgs, funTy, funMeta, funBody} = do 
+   let tailCallTy = markTailCallsFnBody funName ddefs funTy funBody 
+   if elem TC tailCallTy
+   then
+      let (ArrowTy2 locVars arrIns _arrEffs arrOut _locRets _isPar _) = funTy 
+          funTy' = (ArrowTy2 locVars arrIns _arrEffs arrOut _locRets _isPar TC)
+        in dbgTraceIt (sdoc (tailCallTy, funName, funTy')) dbgTraceIt "a" return $ FunDef funName funArgs funTy' funBody funMeta
+   else if elem TMC tailCallTy
+   then 
+      let (ArrowTy2 locVars arrIns _arrEffs arrOut _locRets _isPar _) = funTy
+          funTy' = (ArrowTy2 locVars arrIns _arrEffs arrOut _locRets _isPar TMC)
+        in dbgTraceIt (sdoc (tailCallTy, funName, funTy')) dbgTraceIt "b" return $ FunDef funName funArgs funTy' funBody funMeta
+   else dbgTraceIt (sdoc (tailCallTy, funName, funTy)) dbgTraceIt "c" pure f 
+   --dbgTraceIt (sdoc tailCallTy) pure f
+
+
+markTailCallsFnBody :: Var -> NewL2.DDefs2 -> ArrowTy2 NewL2.Ty2 -> NewL2.Exp2 -> [TailRecType] 
```

Comment:
Judging by how the result of this function is used, I don't think the list type in the result is needed. Instead, the function can use the `Ord` instance of `TailRecType` and always return the `maximum` of the results of subexpressions. You will need to reorder the constructors in the definition of `TailRecType` though. But this will allow you to do away with the nested `if`'s at the use site above, which is a big win: nested `if`'s are better avoided, if fpossible.